File: lua/pascals-triangle/pascals-triangle.lua
local fact = {}

local function factorial(n)
    if n == 1 then return 1 end
    if fact[n] then return fact[n] end
    fact[n] = n * factorial(n-1)
    return fact[n]
end

local function binomial_coefficient(n, k)
    if k == 0 or (n-k) == 0 then return 1 end
    if k == 1 or (n-k) == 1 then return n end
    return factorial(n) / (factorial(k) * factorial(n-k))
end

local function generate_row(n)
    local row = {}
    for k = 1, n do
        row[#row+1] = binomial_coefficient(n-1, k-1)
    end
    return row
end

return function (n)
    local t = {}
    for i = 1, n do
        t[i] = generate_row(i)
    end

    return {
        rows = t,
        last_row = t[n],
    }
end

File: elixir/pascals-triangle/lib/pascals_triangle.ex
defmodule PascalsTriangle do
  @doc """
  Calculates the rows of a pascal triangle
  with the given height
  """
  @spec rows(integer) :: [[integer]]
  def rows(num) do
    do_pascal(num, num, [])
  end

  defp do_pascal(_, 0, result), do: result

  defp do_pascal(num, n, result) do
    do_pascal(num, n - 1, [row(n) | result])
  end

  defp row(n) do
    for k <- 1..n do choose(n - 1, k - 1) end
  end

  # Note: caching the factorial values, although adding complexity,
  # _might_ increase performance. TBD

  defp choose(n, k) do
    div(fact(n), fact(k) * fact(n - k))
  end

  defp fact(0), do: 1
  defp fact(x), do: Enum.reduce(1..x, 1, &*/2)
end
